### ISSUE TITLE
ServerException on sending

### ISSUE THREAD

**Issue by Chuvisco88** [productsupcom/Platform-API-Client#5]
Today we started to get following error:
 
[Productsup\Exceptions\ServerException]                                                                                                       
 file_get_contents(/home/data/api/imports/435122/temp/1977968b17328b21b7c11497c4793b79.csv): failed to open stream: No such file or directory 

Is there any change on your side of the API?

**Comment by 0bp** [comment]
For batch 1977968b17328b21b7c11497c4793b79 we have a commit in our logs but no upload request before.

 That's the cause and indeed we should catch and respond with a meaningful error message.

**Comment by Chuvisco88** [comment]
Should the upload request not happen by the library?
Didn't find anything in the examples about creating an upload request prior to committing.
So far as I have seen we handle it like in the ProductData service from the examples.

**Comment by 0bp** [comment]
Hm, we'll check. There's an automatic commit every 5k(?) products and at the end, when you call the final commit() the product list could be empty because you might have added exactly 5k products. Indeed that should be handled by the client.

**Comment by 0bp** [comment]
We'll investigate the reason and try to reproduce. Internal Ticket is API-26.

> Didn't find anything in the examples about creating an upload request prior to committing.

It would probably happen doing no `->insert(...)` but a `->commit()`

**Comment by Chuvisco88** [comment]
Well we are doing a lot of inserts and a commit in the end when I look at our code. Like you said, there is an auto commit every x product inserts (we use the default value).

Below the code sample how we are calling the api:
```
// $this->service is the product data service from the library
...
    $this->addToBatch($products);
    $this->service->commit();
...
function addToBatch($products) {
    foreach ($products as $product) {
            if ($this->hasDeleteFlag($product)) {
                $this->service->delete($product);
            } else {
                $this->service->insert($product);
            }
     }
}
```

**Comment by Chuvisco88** [comment]
I think I found the bug on in the function _submit (starting from line 222).
For the request the type is not set.

**Comment by 0bp** [comment]
The request type only exists for the /commit endpoint and not for /upload

**Comment by seline196** [comment]
Ok, so we still haven't found the issue? Can someone please take care of it from PU side? We need to solve this issue ASAP. Thank you

**Comment by wernerspiegelpup** [comment]
hi @Chuvisco88, 
can you please sent a more elaborate example of a product and how you sent it to the api.
preferably via mail (ws@productsup.com) or skype (werner.spiegel) 

thank you

**Comment by wernerspiegelpup** [comment]
```
$Client = new Productsup\Client();
$Client->id = '***';
$Client->secret = '***';

$reference = new \Productsup\Platform\Site\Reference();
$reference->setKey(\Productsup\Platform\Site\Reference::REFERENCE_SITE);
$reference->setValue($site_id);

$ProductService = new Productsup\Service\ProductData($Client);
$ProductService->setReference($reference);

/*$ProductService->insert(array(
        'id' => 123,
        'price' => 1.23,
        'description' => 'test title',
    )
);*/

$res = $ProductService->commit();
print_r($res);
```

Fatal error: Uncaught Productsup\Exceptions\ServerException: file_get_contents(/home/data/api/imports/435609/temp/bfa66e4b591341c55d0f015bb82c1e3a.csv): failed to open stream: No such file or directory

with the insert i get:

> Array
> (
>     [0] => 2016-03-31 13:18:23 added 1 product(s)  for upload to batch be8a4581eb7f6b9db142a40c1d9d4210
> )

**Comment by Chuvisco88** [comment]
Interesting point, changed our code so the commit is only done if there are products to commit.
Will keep you posted if the problem occurs again.

**Comment by Chuvisco88** [comment]
@wernerspiegelpup see below for an example product (contained in $products as used in $this->addToBatch($products) from code seen above)
```
array(48) {
  'id' =>
  string(7) "53727-1"
  'merchant_product_id' =>
  string(7) "53727-2"
  'merchant_name' =>
  string(10) "Magando AG"
  'title' =>
  string(45) "Wilkinson - Schick Quattro Titanium Precision"
  'description' =>
  string(168) "- 4 Titan / Carbon beschichtete Klingen<br>- angenehmer Gleitstreifen mit Aloe Vera<br>- Rasierer umdrehen und den Trimmer verwenden für sauberes Konturenschneiden<br>"
  'product_type' =>
  string(54) "Körper & Pflege > Rasieren > Rasierer / Rasierklingen"
  'siroop_category_l1' =>
  string(16) "Körper & Pflege"
  'siroop_category_l2' =>
  string(8) "Rasieren"
  'siroop_category_l3' =>
  string(24) "Rasierer / Rasierklingen"
  'link' =>
  string(122) "https://www.ch.koalaa.dev/koerper-pflege/rasieren/rasierer-rasierklingen/wilkinson-schick-quattro-titanium-precision-53727"
  'image_link' =>
  string(75) "https://cdn.siroop.ch/media/images/sized/LP58Kytv9lFhmft6FMwqgQ.400x400.jpg"
  'condition' =>
  string(3) "new"
  'brand' =>
  string(9) "Wilkinson"
  'availability' =>
  string(8) "in stock"
  'price' =>
  string(5) "16.80"
  'google_product_category' =>
  NULL
  'gtin' =>
  string(12) "841058013006"
  'seo_meta_description' =>
  string(45) "Wilkinson - Schick Quattro Titanium Precision"
  'seo_meta_title' =>
  string(45) "Wilkinson - Schick Quattro Titanium Precision"
  'additional_image_link' =>
  NULL
  'color' =>
  NULL
  'gender' =>
  string(4) "male"
  'age_group' =>
  NULL
  'material' =>
  string(0) ""
  'muster' =>
  string(0) ""
  'size' =>
  NULL
  'size_type' =>
  NULL
  'mpn' =>
  string(0) ""
  'haartyp' =>
  string(0) ""
  'deckkraft' =>
  string(0) ""
  'anschlusstechnik' =>
  string(0) ""
  'digitalkameratyp' =>
  string(0) ""
  'produkttyp' =>
  string(0) ""
  'prozessor' =>
  string(0) ""
  'betriebssystem' =>
  string(0) ""
  'formfaktor' =>
  string(0) ""
  'monitor' =>
  string(0) ""
  'topseller_l1' =>
  string(0) ""
  'topseller_l2' =>
  string(0) ""
  'topseller_l3' =>
  string(0) ""
  'topseller_home' =>
  string(0) ""
  'featured_product_l1' =>
  string(0) ""
  'product_score' =>
  int(0)
  'tags' =>
  string(122) "Körper & Pflege,Rasieren,Rasierer / Rasierklingen,Gesundheit & Schönheit,Körperpflege,Rasieren & Trimmen,Haarentfernung"
  'lowest_price' =>
  int(1)
  'currency' =>
  string(3) "CHF"
  'is_active' =>
  bool(true)
  'is_active_abstract' =>
  bool(true)
}
```

**Comment by wernerspiegelpup** [comment]
i'll throw client exception if no products have been submitted on a commit() call and also throw a meaningful server exception if this occurs and notify you when this is live.

thanks for bughunting :+1:

**Comment by wernerspiegelpup** [comment]
released [2.0.0-RC2](https://github.com/productsupcom/Platform-API-Client/tree/2.0.0-RC2)
